classes/local/model/error.php

```diff
@@ -0,0 +1,200 @@
+    const STATUS_PENDING = 'pending';
+
+    /** For errors currently being processed. */
+    const STATUS_RETRYING = 'retrying';
+
+    /** Successfully re-indexed. */
+    const STATUS_RESOLVED = 'resolved';
+
+    /** Exceeded the max retry limit. */
+    const STATUS_FAILED = 'failed';
+
+    /** Content no longer exists. */
+    const STATUS_OBSOLETE = 'obsolete';
+
+    /** @var int The maximum retry attempts. */
+    const MAX_RETRIES = 10;
+
+    /**
+     * Return the definition of the properties of this model.
+     *
+     * @return array
+     */
+    protected static function define_properties() {
+        return [
+            'docid' => [
+                'type' => PARAM_TEXT,
+                'null' => NULL_NOT_ALLOWED,
+            ],
+            'itemid' => [
+                'type' => PARAM_INT,
+                'null' => NULL_NOT_ALLOWED,
+            ],
+            'contextid' => [
+                'type' => PARAM_INT,
+                'null' => NULL_ALLOWED,
+                'default' => null,
+            ],
+            'areaid' => [
+                'type' => PARAM_TEXT,
+                'null' => NULL_NOT_ALLOWED,
+            ],
+            'errortype' => [
+                'type' => PARAM_ALPHA,
+                'null' => NULL_NOT_ALLOWED,
+                'choices' => [
+                    self::TYPE_INDEXING,
+                    self::TYPE_TIKA,
+                ],
+            ],
+            'errormessage' => [
+                'type' => PARAM_TEXT,
+                'null' => NULL_NOT_ALLOWED,
+            ],
+            'retrycount' => [
+                'type' => PARAM_INT,
+                'null' => NULL_NOT_ALLOWED,
+                'default' => 0,
+            ],
+            'status' => [
+                'type' => PARAM_TEXT,
+                'null' => NULL_NOT_ALLOWED,
+                'default' => self::STATUS_PENDING,
+                'choices' => [
+                    self::STATUS_PENDING,
+                    self::STATUS_RETRYING,
+                    self::STATUS_RESOLVED,
+                    self::STATUS_FAILED,
+                    self::STATUS_OBSOLETE,
+                ],
+            ],
+            'timecreated' => [
+                'type' => PARAM_INT,
+                'null' => NULL_NOT_ALLOWED,
+            ],
+            'timemodified' => [
+                'type' => PARAM_INT,
+                'null' => NULL_NOT_ALLOWED,
+            ],
+            'contentmodified' => [
+                'type' => PARAM_INT,
+                'null' => NULL_ALLOWED,
+                'default' => null,
+            ],
+            'parentid' => [
+                'type' => PARAM_TEXT,
+                'null' => NULL_ALLOWED,
+                'default' => null,
+            ],
+        ];
+    }
+
+    /**
+     * Hook to execute before the record is created.
+     */
+    protected function before_create() {
+        $now = time();
+        $this->set('timecreated', $now);
```

Comment:
I was under impression that this is managed in persistent class. Why do we need manage timecreated and timemodified in our class?